handbook/engineering/testing.md

```diff
@@ -2,11 +2,87 @@
 
 This file documents how we test code at Sourcegraph.
 
+## Philosophy
 
-## Percy
+We rely on automated testing to ensure the quality of our product.
+
+Any addition or change to our codebase should be covered by an appropriate amount of automated tests to ensure that:
+
+1. Our product works as intended when we ship it to customers.
+1. Our product continues to work and doesn't accidentally break as we make changes to our product over time.
+
+Engineers should budget for writing tests when making iteration plans.
+
+## Flaky tests
+
+We do not tolerate flaky tests of any kind. Any engineer that sees a flaky test should immediately:
+
+1. Open a PR to disable the flaky test.
+1. Open an issue to re-enable the flaky test, assign it to the most likely owner, and add it to the current release milestone.
+
+If the build or test infrastructure itself is flaky, then [open an issue](https://github.com/sourcegraph/sourcegraph/issues/new?labels=team/distribution) and notify the [distribution team](distribution/index.md#contact).
+
+## Unit tests
+
+Unit tests test individual functions in our codebase and are the most desirable kind of test to write.
+
+Benefits:
+
+- They are usually very fast to execute because slow operations can be mocked.
+- They are the easiest tests to write, debug, and maintain because the code under test is small.
+- They only need to run on changes that touch code which could make the test fail, which makes CI faster and minimizes the impact of any flakiness.
+
+Tradeoffs:
+
+- They don't verify our systems are wired up correctly end-to-end.
```

Comment:
I think this trade-off should be called out - I frequently meet this in the webapp. This is better or worse depending on how well designed the implementation and test is, but unit tests are also generally more vulnerable to this because they test the smallest unit of test, and any change that removes the unit, splits the unit, merges two units, changes the return type of the unit, etc etc will require updates to unit tests (while integration tests and e2e tests are more likely to remain unchanged). This can result in _a lot_ of work per change and can become a change preventer.

```suggestion
- They don't verify our systems are wired up correctly end-to-end.
- They often require changes when refactoring code.
```

Reply:
> any change that removes the unit, splits the unit, merges two units, changes the return type of the unit, etc etc will require updates to unit tests... This can result in a lot of work per change and can become a change preventer.

My feeling is that if the refactoring changes described here impact tests, then changing the associated unit tests is also a benefit. I'm not saying that's not a lot of work (there is a tradeoff), but this behavior is also beneficial in that if the changes result in a lot of work, that only speaks to the significance of the change in structure in question that we should take care to keep testing.

Reply:
I share Rijnard's perspective here: I don't think this is a tradeoff of unit tests (as compared to integration or e2e tests). You don't experience this with e2e tests because it simply isn't possible or feasible to test the entire state space of what you want/should test. If you did/could, then e2e tests would have the same problem.

To the extent that state space of the code changes, then yes, tests need to be updated to cover that new state space (because testing is valuable), and that is a good thing (not a bad thing).

Writing good tests is extra work, and the point of adding this to our handbook is to document that engineers should do this work. The extra work bit is not a tradeoff of unit tests in particular, it is a feature of us prioritizing testing at all and unit tests being the most effective way to test most of the state space.